assets/js/blocks/products/utils.js

```diff
@@ -11,14 +11,39 @@ export const getBlockClassName = ( blockClassName, attributes ) => {
 	const { className, contentVisibility } = attributes;
 
 	return classNames( blockClassName, className, {
-		'has-image': contentVisibility.image,
-		'has-title': contentVisibility.title,
-		'has-rating': contentVisibility.rating,
-		'has-price': contentVisibility.price,
-		'has-button': contentVisibility.button,
+		'has-image': contentVisibility && contentVisibility.image,
+		'has-title': contentVisibility && contentVisibility.title,
+		'has-rating': contentVisibility && contentVisibility.rating,
+		'has-price': contentVisibility && contentVisibility.price,
+		'has-button': contentVisibility && contentVisibility.button,
 	} );
 };
 
+/**
+ * Adjust attributes on load to set defaults so default attributes get saved.
+ *
+ * Ref: https://github.com/WordPress/gutenberg/issues/7342
+ *
+ * @param {Object} blockAttributes Original block attributes.
+ * @param {Object} blockType       Block type settings.
+ *
+ * @return {Object} Filtered block attributes.
+ */
+export const setBlockAttributeDefaults = ( blockAttributes, blockType ) => {
+	Object.keys( blockType.attributes ).map( ( key ) => {
+		if (
+			blockAttributes[ key ] === undefined &&
+			blockType.defaults !== undefined &&
+			blockType.defaults[ key ] !== undefined
+		) {
+			blockAttributes[ key ] = blockType.defaults[ key ];
+		}
```

Comment:
I wonder if instead of direct equality we should use `typeof` checks instead.  This doesn't have a chance (qualifier: assuming `blockAttributes` and `blockType` are always defined!)  of throwing a `ReferenceError` in this case because you are checking for properties on an object, however it's a bad habit to check for undefined in this way because undeclared references _will_ throw errors.

This comment applies to the other places you are doing this same type of check in this pull.